doc: add fault resolution ADR (#2051)
cosmos/interchain-security: docs/docs/adrs/adr-018-fault-resolutions.md

 This threat is particularly relevant for Opt-in chains since they might be secured by a relatively small subset of the provider's validator set.  

In cases of collusion, various types of misbehaviour can be performed by the validators, such as:

* Incorrect executions to break protocol rules in order to steal funds.
* Liveness attacks to halt the chain or censor transactions.
* Oracle attacks to falsify information used by the chain logic.

Currently, these types of attacks aren't handled in PSS, leaving the malicious validators unpunished.

A potential solution is to use fraud proofs. This technology allows proving incorrect state transitions of a chain without a full node.
 However, this is a complex technology, and there is no framework that works for Cosmos chains to this day.

To address this risk in PSS, a governance-gated slashing solution can be used until fraud proof technology matures.

This ADR proposes a fault resolution mechanism, which is a type of governance proposal that victims of faults can use to vote on the
 slashing of validators that misbehave on Opt-in consumer chains.

> In what follows, we describe the implementation of a fault resolution mechanism that handles incorrect executions on consumer chains,

Author: Because we only provide a guideline for incorrect execution fault definition in the intersubjective faults CHIPs. The handling for other types of faults will be implemented iteratively.

> In what follows, we describe the implementation of a fault resolution mechanism that handles incorrect executions on consumer chains,

Reviewer: but that will not need any new implementation effort, right?

> In what follows, we describe the implementation of a fault resolution mechanism that handles incorrect executions on consumer chains,

Reviewer: I would recommend making this a bit more general. From a code perspective, this handles arbitrary fault resolutions

> In what follows, we describe the implementation of a fault resolution mechanism that handles incorrect executions on consumer chains,

Author: As we all agree the ADR is specific to incorrect executions and that should be reflected in its title.
The filename will be renamed in the last step so as not to confuse the review process.

 as a first iteration.

## Decision

The proposed solution introduces a new `consumer-fault-resolution` governance proposal type to the `provider` module, which allows
 validators to be penalised for committing faults on an Opt-in consumer chain.

If such a proposal passes, the proposal handler tombstones all the validators listed in the proposal and slashes them by a predefined
 amount or the default value used for double-sign infractions.

The proposal has the following fields:

- **Description**: This field should be filled with a fault definition describing the type of misbehavior that the validators executed
 on an Opt-in consumer chain. A fault definition should precisely describe how an attack was performed and why it is eligible as a slashable fault.
- **Consumer Chain**: The chain that the fault was related to.

> - **Validators**: The list of all the validators to be slashed.

Reviewer 2: @mpoke I disagree. @sainoe add an "evidence" text field

In addition, in order to prevent spamming, users are required to pay a fee of `100ATOM` to submit a fault resolution to the provider.

### Validations

> The submission of a fault resolution fails if any of the following conditions are not met:

Reviewer 3: Avoid the double negation
```suggestion
The submission of a fault resolution succeeds only if all of the following conditions are met:
```

- the consumer chain is an Opt-in chain

> - all listed validators were opted-in to the consumer chain in the past unbonding-period

Reviewer 3: ```suggestion
- the provided validator set matches one of the consumer validator sets from the previous unbonding period
- all the listed validators are part of the provided validator set  
```

Reviewer 2: @sainoe yes this is right. Perhaps only a few validators double signed.

- the `100ATOM` fee is provided

### Additional considerations

Fault resolution proposals should be `expedited`  to minimize the time given to the listed validators to
to unbond to avoid punishment (see [Expedited Proposals](https://docs.cosmos.network/v0.50/build/modules/gov#expedited-proposals)) .

## Consequences

### Positive

- Provide the ability to slash and tombstone validators for committing incorrect executions on Opt-in consumer chains.

### Negative

- Assuming that malicious validators unbond immediately after misbehaving, a fault resolution has to be submitted within a maximum
 of two weeks in order to slash the validators.

### Neutral

- Fault definitions need to have a clear framework in order to avoid debates about whether an attack has actually taken place.  

> ## References

Reviewer 3: Add reference to https://forum.cosmos.network/t/preventing-intersubjective-faults-in-ics/14103

 However, this is a complex technology and and there is no framework that works for Cosmos chains to this day.

To address this risk in PSS, a governance-gated slashing solution can be used to handle all types of misbehavior resulting from validator collusion. As fraud proof technology matures, part of the solution could potentially be automated.

This ADR proposes a fault resolution mechanism, which is a type of governance proposal that can be used to vote on the slashing of validators that misbehave on Opt-in consumer chains.

In what follows, we describe the implementation of a fault resolution mechanism that handles incorrect executions on consumer chains,
 as a first iteration.

## Decision

The proposed solution introduces a new `consumer-fault-resolution` governance proposal type to the `provider` module, which allows
 validators to be penalised for committing faults on an Opt-in consumer chain.

If such a proposal passes, the proposal handler tombstones all the validators listed in the proposal and slashes them by a per-consumer chain predefined
 amount or the default value used for double-sign infractions.

The proposal has the following fields:

> - **Description**: This field should be filled with a fault definition describing the type of misbehavior that the validators executed

Reviewer 2: Just to be clear- the user should not fill in the fault definition. It should be included in the prop automatically.
